File: VB/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalWeekOfYearFunction.vb
Imports System
Imports System.Globalization
Imports System.Linq

Namespace Dashboard_FiscalFunctions
	Friend Class GetFiscalWeekOfYearFunction
		Inherits GetFiscalDateFunction

		Private rule As CalendarWeekRule
		Private firstDay As DayOfWeek

		Public Sub New(ByVal startDay As Integer, ByVal startMonth As Integer, ByVal rule As CalendarWeekRule, ByVal firstDay As DayOfWeek)
			MyBase.New(startDay, startMonth)
			Me.rule = rule
			Me.firstDay = firstDay
		End Sub

		Public Overrides ReadOnly Property Name() As String
			Get
				Return "GetFiscalWeekOfYear"
			End Get
		End Property

		Public Overrides ReadOnly Property Description() As String
			Get
				Return "GetFiscalWeek(DateTime) " & vbLf & " Extracts a fiscal week of the year from the defined DateTime"
			End Get
		End Property

		Private Function WeekCounts(ByVal current As Date, ByVal difference As Integer) As Integer
			For i As Integer = 1 To difference - 1
				If (CInt(current.DayOfWeek) + i + 7) Mod 7 = CInt(firstDay) Then
					Return i
				End If
			Next i
			Return -1
		End Function

		Private Function Beginning(ByVal currentYear As Integer) As Date
			Dim start As New Date(currentYear, startMonth, startDay)
			If rule = CalendarWeekRule.FirstDay Then
				Return start
			Else
				Dim range As Integer
				If rule = CalendarWeekRule.FirstFourDayWeek Then
					range = WeekCounts(start, 4)
				Else
					range = WeekCounts(start, 7)
				End If
				If range = -1 Then
					Return start
				Else
					Return start.AddDays(range)
				End If
			End If
		End Function

		Protected Overrides Function GetFiscal(ByVal current As Date) As Integer
			Dim start As Date = Beginning(current.Year)
			If current < start Then
				start = Beginning(current.Year - 1)
			End If
			Dim plus As Integer = (start.DayOfWeek - firstDay + 7) Mod 7
			Return (CInt(Math.Truncate((current.Subtract(start)).TotalDays)) + plus) \ 7 + 1
		End Function
	End Class
End Namespace

File: CS/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalWeekOfYearFunction.cs
using System;
using System.Globalization;
using System.Linq;

namespace Dashboard_FiscalFunctions {
    class GetFiscalWeekOfYearFunction : GetFiscalDateFunction {
        CalendarWeekRule rule;
        DayOfWeek firstDay;

        public GetFiscalWeekOfYearFunction(int startDay, int startMonth, CalendarWeekRule rule, DayOfWeek firstDay) : base(startDay, startMonth) {
            this.rule = rule;
            this.firstDay = firstDay;
        }

        public override string Name => "GetFiscalWeekOfYear";

        public override string Description => "GetFiscalWeek(DateTime) \n Extracts a fiscal week of the year from the defined DateTime";

        int WeekCounts(DateTime current, int difference) {
            for(int i = 1; i < difference; i++) {
                if(((int)current.DayOfWeek + i + 7) % 7 == (int)firstDay) {
                    return i;
                }
            }
            return -1;
        }

        DateTime Beginning(int currentYear) {
            DateTime start = new DateTime(currentYear, startMonth, startDay);
            if(rule == CalendarWeekRule.FirstDay) {
                return start;
            } else {
                int range;
                if(rule == CalendarWeekRule.FirstFourDayWeek) {
                    range = WeekCounts(start, 4);
                } else {
                    range = WeekCounts(start, 7);
                }
                if(range == -1) {
                    return start;
                } else {
                    return start.AddDays(range);
                }
            }
        }

        protected override int GetFiscal(DateTime current) {
            DateTime start = Beginning(current.Year);
            if(current < start) {
                start = Beginning(current.Year - 1);
            }
            int plus = (start.DayOfWeek - firstDay + 7) % 7;
            return ((int)(current - start).TotalDays + plus) / 7 + 1;
        }
    }
}
